### PR TITLE
[dynamo] fix functools.wraps on nested functions

### LINKED ISSUE #103913: [dynamo] functools.wraps : graph-break when wrapping nested functions.
### 🐛 Describe the bug

Repro:
```python
import torch
from torch._dynamo.utils import counters
import json
from functools import wraps

counters.clear()

def foo(x, y):
    def add(x, y):
        return x + y

    @wraps(add)
    def wrapped_call(x, y):
        return add(x, y)

    return wrapped_call(x, y)

x = torch.ones(1,)
y = torch.ones(1,)

o = torch.compile(foo, fullgraph=False, backend='aot_eager')(x, y)

torch.testing.assert_close(o, 2 * x)
assert counters["graph_break"] == {
    "call_function wraps in skip_files /home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.9/functools.py": 1,
    "call_method UserDefinedObjectVariable(partial) __call__ [NestedUserFunctionVariable()] {}": 1
}
# print(json.dumps(counters, indent=2))

```
Note, that it doesn't graph-break for non-nested functions
```python
import torch
from torch._dynamo.utils import counters
import json
from functools import wraps

counters.clear()

def add(x, y):
        return x + y

def foo(x, y):
    @wraps(add)
    def wrapped_call(x, y):
        return add(x, y)

    return wrapped_call(x, y)

x = torch.ones(1,)
y = torch.ones(1,)

o = torch.compile(foo, fullgraph=False, backend='aot_eager')(x, y)

torch.testing.assert_close(o, 2 * x)
assert counters["graph_break"] == {}
# print(json.dumps(counters, indent=2))
```

Found in https://github.com/pytorch/pytorch/pull/102264 (this leads to graph-breaks when using nested functions with `torch.func.grad`)

### Versions

master

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @wconstab @Xia-Weiwen @ipiszy

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes https://github.com/pytorch/pytorch/issues/103913

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: test/functorch/test_control_flow.py [hunk 1]
```diff
@@ -1422,11 +1422,21 @@ def wrapper(*args, **kwargs):
...
+        # The expected graph differs slight when running under Dynamo
+        if TEST_WITH_TORCHDYNAMO:
+            fwd_string = "def forward(self, arg0_1, arg1_1):"
+            impl_string = ("map_impl = torch.ops.map_impl(body_graph_0, 1, arg1_1, arg0_1);"
+                           "body_graph_0 = arg1_1 = arg0_1 = None")
+        else:
+            fwd_string = "def forward(self, pred_1, x_1):"
+            impl_string = ("map_impl = torch.ops.map_impl(body_graph_0, 1, x_1, pred_1);"
+                           "body_graph_0 = x_1 = pred_1 = None")
+
```
Comment (zou3519) [inline test/functorch/test_control_flow.py:1435]: This seems unexpected, do we know what's going on here? Dynamo is supposed to faithfully replicate what happens in eager mode

File: torch/_dynamo/variables/misc.py [hunk 2]
```diff
@@ -767,16 +767,15 @@ def func(tx, item, acc):
...
+        elif self.value is functools.wraps and not kwargs and len(args) == 1:
 
             def wraps(fn):
                 if isinstance(fn, variables.NestedUserFunctionVariable):
-                    return fn.clone(wraps_source=args[0].source)
+                    # `wraps_source` should implement `reconstruct`
+                    # which NestedUserFunctionVariable does.
+                    # We prefer `args[0].source` if that is available.
+                    wraps_source = args[0].source if args[0].source else fn
+                    return fn.clone(wraps_source=wraps_source)
```
Comment (zou3519) [inline torch/_dynamo/variables/misc.py:778]: @kshitij12345 IMO we should always just pass one of them to be consistent. Do things break if we always pass NestedUserFunctionVariable?

Comment (zou3519) [inline torch/_dynamo/variables/misc.py:778]: @voznesenskym could we get a sanity check here please?
- Previously, functools.wraps(f) support didn't work if f (a NestedUserFunctionVariable) had no source. (NestedUserFunctionVariable created in the current Dynamo context don't have a source.)
- As far as we can tell, f's source is used to reconstruct the `functools.wraps(f)` call during a graph break: the codegen uses the source of `f` to pull `f`, and then invokes functools.wraps on it.
- The new strategy is to reconstruct the `functools.wraps(f)` call by literally calling `f.reconstruct()` (since every NestedUserFunctionVariable has a reconstruct() method) and then calling functools.wraps on it.

File: test/functorch/test_control_flow.py (hunk 1 above)
Comment (kshitij12345) [reply test/functorch/test_control_flow.py]: That was a bug, latest commit fixes it. Thanks! (I initially thought it was ok since just the argument names changed)

File: torch/_dynamo/variables/misc.py (hunk 2 above)
Comment (kshitij12345) [reply torch/_dynamo/variables/misc.py]: Have made an update, passing `fn` was the wrong thing to do.

File: torch/_dynamo/variables/misc.py [hunk 3]
```diff
@@ -767,16 +767,20 @@ def func(tx, item, acc):
...
+                # Eg.
+                #   wraps(foo)
+                #   def bar():
+                #      pass
+                # Here : args[0] -> foo, fn -> bar
                 if isinstance(fn, variables.NestedUserFunctionVariable):
-                    return fn.clone(wraps_source=args[0].source)
+                    # `wraps_source` should either have a source
+                    # or implement reconstruct.
+                    return fn.clone(wraps_source=args[0])
```
Comment (voznesenskym) [inline torch/_dynamo/variables/misc.py:783]: Lol, this works but let's not do this. You are taking advantage of the fact that wraps_source is only used in reconstruct. That works until someone tries to access this, thinking its a source, but then its a function! This is how frameworks degrade, imo.

Instead:

1) add a `can_reconstruct` util to base VT 
2) rename `wraps_source` to `reconstructable` and type it to a union of Source and VT
3) maintain the invariant that `codegen` on the NestedUserFunctionVariable w/ a `reconstructable` can reconstruct

**Review by voznesenskym — CHANGES_REQUESTED** [review]
See feedback

File: torch/_dynamo/variables/misc.py (hunk 3 above)
Comment (kshitij12345) [reply torch/_dynamo/variables/misc.py:783]: > (wraps_source is only used for functools.wraps, and the input is always a NestedUserFunctionVariable?)

`wraps_source` is used only for functools.wraps is correct. However input may or may not be a `NestedUserFunctionVariable` (it can also be `UserFunctionVariable` when the wrapped function is in global namespace).

`NestedUserFunctionVariable` itself implements reconstruct (which codegen uses), 
https://github.com/pytorch/pytorch/blob/bc3f0d341aade5f6e444b39fc3c6ee7a8e49d2cf/torch/_dynamo/codegen.py#L147-L155

while in case of `UserFunctionVariable`, it has source and codegen knows how to deal with that.
https://github.com/pytorch/pytorch/blob/bc3f0d341aade5f6e444b39fc3c6ee7a8e49d2cf/torch/_dynamo/codegen.py#L97-L102



Ref:
Usage of `wraps_source`
https://github.com/pytorch/pytorch/blob/bc3f0d341aade5f6e444b39fc3c6ee7a8e49d2cf/torch/_dynamo/variables/functions.py#L524

NOTE: codegen doesn't seem to be typed and deals with a variety of types
https://github.com/pytorch/pytorch/blob/bc3f0d341aade5f6e444b39fc3c6ee7a8e49d2cf/torch/_dynamo/codegen.py#L70-L157

---------

I think renaming `wraps_source` to `wraps_func` makes sense, as it is the function being wrapped. We can type it as `BaseUserFunctionVariable` or  if we want to play safer as `Union[UserFunctionVariable, NestedUserFunctionVariable]`.

Wdyt?

Thanks!

Comment (zou3519) [reply torch/_dynamo/variables/misc.py:783]: On second thought, people can call functools.wraps on anything, right? (it does not need to be a UserFunctionVariable/NestedUserFunctionVariable). If that's the case, then we would want to apply Voz's suggestion of having a `wraps_fun_reconstructible` that is Union[Source, UserFunctionVariable, NestedUserFunctionVariable]

Comment (kshitij12345) [reply torch/_dynamo/variables/misc.py:783]: @voznesenskym 

Regarding the `can_reconstruct`, I don't think we can easily figure that for a subclass of VariableTracker.

We can't implement it like `can_make_guard` as `reconstruct` expects a codegen object and also codegen mutates the translator when called.
https://github.com/pytorch/pytorch/blob/7f7f6267e99f8dd5990e43dd2301ac37bb66e299/torch/_dynamo/variables/base.py#L241-L246
https://github.com/pytorch/pytorch/blob/7f7f6267e99f8dd5990e43dd2301ac37bb66e299/torch/_dynamo/variables/base.py#L284-L285

Another concern I have is about the notion of reconstructable VT. Codegen doesn't always call `reconstruct` on VariableTracker but handles them case by case (without calling `reconstruct`). So even if we implement `can_reconstruct` like above, it wouldn't still cover all the tracker subclasses that codegen can deal with.
https://github.com/pytorch/pytorch/blob/7f7f6267e99f8dd5990e43dd2301ac37bb66e299/torch/_dynamo/codegen.py#L97-L137

So, with above in mind, one approach could be just adding a flag in VT and then all subclasses have to set it manually. (If we opt for this, I think it should be done in separate PR)

For this PR would it be ok to rename `wraps_source` to `wraps_reconstructible : Union[Source, NestedUserFunctionVariable]` and graph break for other cases (and follow-up separately on reconstruct)?

What do you think about this? Is there anything missing?

Thanks!

Comment (kshitij12345) [reply torch/_dynamo/variables/misc.py:783]: Gentle ping @voznesenskym :)

Comment (kshitij12345) [reply torch/_dynamo/variables/misc.py:783]: Gentle ping @voznesenskym
